File: src/java/src/mil/navy/nrl/norm/io/NormInputStream.java
package mil.navy.nrl.norm.io;

import java.io.IOException;
import java.io.InputStream;
import java.net.SocketException;
import java.util.LinkedList;
import java.util.List;

import mil.navy.nrl.norm.NormEvent;
import mil.navy.nrl.norm.NormInstance;
import mil.navy.nrl.norm.NormNode;
import mil.navy.nrl.norm.NormObject;
import mil.navy.nrl.norm.NormSession;
import mil.navy.nrl.norm.NormStream;
import mil.navy.nrl.norm.enums.NormEventType;
import mil.navy.nrl.norm.enums.NormObjectType;

/**
 * @author Jason Rush
 */
public class NormInputStream extends InputStream {
  private NormInstance normInstance;
  private NormSession normSession;
  private NormStream normStream;

  private List<INormEventListener> normEventListeners;

  private boolean closed;
  private Object closeLock;

  private boolean bufferIsEmpty;
  private boolean receivedEof;

  public NormInputStream(String address, int port) throws IOException {
    // Create the NORM instance
    normInstance = new NormInstance();

    // Create the NORM session
    normSession = normInstance.createSession(address, port,
        NormNode.NORM_NODE_ANY);

    normStream = null;

    normEventListeners = new LinkedList<INormEventListener>();

    closed = true;
    closeLock = new Object();

    bufferIsEmpty = true;
    receivedEof = false;
  }

  public synchronized void openDebugLog(String filename) throws IOException {
    normInstance.openDebugLog(filename);
  }

  public synchronized void closeDebugLog() {
    normInstance.closeDebugLog();
  }

  public synchronized void setDebugLevel(int level) {
    normInstance.setDebugLevel(level);
  }

  public synchronized void setMessageTrace(boolean messageTrace) {
    normSession.setMessageTrace(messageTrace);
  }

  public synchronized void setMulticastInterface(String multicastInterface)
      throws IOException {
    normSession.setMulticastInterface(multicastInterface);
  }

  public synchronized void setEcnSupport(boolean ecnEnable, boolean ignoreLoss)
      throws IOException {
    normSession.setEcnSupport(ecnEnable, ignoreLoss);
  }

  public synchronized void setTtl(byte ttl) throws IOException {
    normSession.setTTL(ttl);
  }

  public synchronized void setTos(byte tos) throws IOException {
    normSession.setTOS(tos);
  }

  public synchronized void setSilentReceiver(boolean silent, int maxDelay) {
    normSession.setSilentReceiver(silent, maxDelay);
  }

  public synchronized void setDefaultUnicastNack(boolean defaultUnicastNack) {
    normSession.setDefaultUnicastNack(defaultUnicastNack);
  }

  public synchronized void seekMsgStart() {
    if (normStream == null) {
      throw new IllegalStateException(
          "Can only seek msg start after the stream is connected");
    }

    normStream.seekMsgStart();
  }

  /**
   * @param normEventListener The INormEventListener to add.
   */
  public void addNormEventListener(INormEventListener normEventListener) {
    synchronized (normEventListeners) {
      normEventListeners.add(normEventListener);
    }
  }

  /**
   * @param normEventListener The INormEventListener to remove.
   */
  public void removeNormEventListener(INormEventListener normEventListener) {
    synchronized (normEventListeners) {
      normEventListeners.remove(normEventListener);
    }
  }

  private void fireNormEventOccured(NormEvent normEvent) {
    synchronized (normEventListeners) {
      for (INormEventListener normEventListener : normEventListeners) {
        normEventListener.normEventOccurred(normEvent);
      }
    }
  }

  public void open(long bufferSpace) throws IOException {
    synchronized (closeLock) {
      if (!isClosed()) {
        throw new SocketException("Stream is already open");
      }

      normSession.startReceiver(bufferSpace);

      closed = false;
    }
  }

  /**
   * @see java.io.InputStream#close()
   */
  @Override
  public void close() throws IOException {
    synchronized (closeLock) {
      if (isClosed()) {
        return;
      }

      normSession.stopReceiver();
      normInstance.stopInstance();

      closed = true;
    }
  }

  /**
   * @return Returns the closed.
   */
  public boolean isClosed() {
    synchronized (closeLock) {
      return closed;
    }
  }

  /**
   * @see java.io.InputStream#read()
   */
  @Override
  public synchronized int read() throws IOException {
    byte buffer[] = new byte[1];

    if (isClosed()) {
      throw new IOException("Stream is closed");
    }

    if (read(buffer) < 0) {
      return -1;
    }

    return buffer[0];
  }

  /**
   * @see java.io.InputStream#read(byte[], int, int)
   */
  @Override
  public synchronized int read(byte[] buffer, int offset, int length)
      throws IOException {
    int n;

    if (isClosed()) {
      throw new IOException("Stream is closed");
    }

    do {
      while (bufferIsEmpty || normInstance.hasNextEvent(0, 0)) {
        processEvent();

        if (receivedEof) {
          return -1;
        }
      }

      if (normStream == null) {
        return -1;
      }

      // Read from the stream
      if ((n = normStream.read(buffer, offset, length)) < 0) {
        throw new StreamBreakException("Break in stream integrity");
      }

      bufferIsEmpty = (n == 0);
    } while (bufferIsEmpty);

    return n;
  }

  private void processEvent() throws IOException {
    // Retrieve the next event
    NormEvent normEvent = normInstance.getNextEvent();

    // Check if the stream was closed
    if (isClosed()) {
      throw new IOException("Stream closed");
    }

    if (normEvent != null) {
      // Process the event
      NormEventType eventType = normEvent.getType();
      switch (eventType) {
        case NORM_RX_OBJECT_NEW:
          NormObject normObject = normEvent.getObject();
          if (normObject.getType() == NormObjectType.NORM_OBJECT_STREAM) {
            normStream = (NormStream)normObject;
          }
          break;

        case NORM_RX_OBJECT_UPDATED:
          NormObject object = normEvent.getObject();
          if (!object.equals(normStream)) {
            break;
          }

          // Signal that the buffer is not empty
          bufferIsEmpty = false;
          break;

        case NORM_RX_OBJECT_ABORTED:
        case NORM_RX_OBJECT_COMPLETED:
          normStream = null;

          // Signal that the stream has ended
          receivedEof = true;
          break;

        default:
          break;
      }

      // Notify listeners of the norm event
      fireNormEventOccured(normEvent);
    }
  }
}

File: src/dotnet/src/Mil.Navy.Nrl.Norm/IO/NormInputStream.cs
using System.Runtime.CompilerServices;

namespace Mil.Navy.Nrl.Norm.IO
{
    public class NormInputStream : Stream
    {
        private NormInstance _normInstance;
        private NormSession _normSession;
        private NormStream? _normStream;

        private List<INormEventListener> _normEventListeners;

        private bool _closed;
        private object _closeLock;

        private bool _bufferIsEmpty;
        private bool _receivedEof;

        /// <exception cref="IOException"></exception>
        public NormInputStream(string address, int port)
        {
            // Create the NORM instance
            _normInstance = new NormInstance();

            // Create the NORM session
            _normSession = _normInstance.CreateSession(address, port, NormNode.NORM_NODE_ANY);
            _normStream = null;

            _normEventListeners = new List<INormEventListener>();

            _closed = true;
            _closeLock = new object();

            _bufferIsEmpty = true;
            _receivedEof = false;
        }

        /// <exception cref="IOException"></exception>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void OpenDebugLog(string fileName)
        {
            if (fileName == null)
            {
                throw new IOException("File was name was not found.");
            }
            _normInstance.OpenDebugLog(fileName);
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public void CloseDebugLog() => _normInstance.CloseDebugLog();

        [MethodImpl(MethodImplOptions.Synchronized)]
        public void NormSetDebugLevel(int level) { _normInstance.DebugLevel = level; }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public void SetMessageTrace(bool messageTrace) => _normSession.SetMessageTrace(messageTrace);

        /// <exception cref="IOException"></exception>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void SetMulticastInterface(string multicastInterface)
        {
            _normSession.SetMulticastInterface(multicastInterface);
        }

        /// <exception cref="IOException"></exception>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void SetEcnSupport(bool ecnEnable, bool ignoreLoss)
        {
            _normSession.SetEcnSupport(ecnEnable, ignoreLoss);
        }

        /// <exception cref="IOException"></exception>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void SetTtl(byte ttl)
        {
            _normSession.SetTTL(ttl);
        }

        /// <exception cref="IOException"></exception>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void SetTos(byte tos)
        {
            _normSession.SetTOS(tos);
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public void setSilentReceiver(bool silent, int maxDelay)
        {
            _normSession.SetSilentReceiver(silent, maxDelay);
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public void SetDefaultUnicastNack(bool defaultUnicastNack)
        {
            _normSession.SetDefaultUnicastNack(defaultUnicastNack);
        }

        /// <exception cref="InvalidOperationException"></exception>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void SeekMsgStart()
        {
            if (_normStream == null)
            {
                throw new InvalidOperationException("Can only seek msg start after the stream is connected");
            }

            _normStream.SeekMsgStart();
        }

        /// <param name="normEventListener">The INormEventListener to add.</param>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void AddNormEventListener(INormEventListener normEventListener)
        {
            lock (_normEventListeners)
            {
                _normEventListeners.Add(normEventListener);
            }
        }

        /// <param name="normEventListener">The INormEventListener to remove.</param>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void RemoveNormEventListener(INormEventListener normEventListener)
        {
            lock (_normEventListeners)
            {
                _normEventListeners.Remove(normEventListener);
            }
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        private void FireNormEventOccured(NormEvent normEvent)
        {
            lock (_normEventListeners)
            {
                foreach (var normEventListener in _normEventListeners)
                {
                    normEventListener.NormEventOccurred(normEvent);
                }

            }
        }

        /// <exception cref="IOException"></exception>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void Open(long bufferSpace)
        {
            lock (_closeLock)
            {
                if (!IsClosed)
                {
                    throw new IOException("Stream is already open");
                }

                _normSession.StartReceiver(bufferSpace);

                _closed = false;
            }
        }

        /// <exception cref="IOException"></exception>
        public override void Close()
        {
            lock (_closeLock)
            {
                if (IsClosed)
                {
                    return;
                }

                _normStream?.Close(false);
                _normSession.StopSender();
                _normInstance.StopInstance();

                _closed = true;
            }
        }

        public bool IsClosed
        {
            get
            {
                lock (_closeLock)
                {
                    return _closed;
                }
            }
        }

        /// <exception cref="IOException"></exception>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public int Read()
        {
            var buffer = new byte[1];

            if (IsClosed)
            {
                throw new IOException("Stream is closed");
            }

            if (Read(buffer) < 0)
            {
                return -1;
            }

            return buffer[0];
        }

        /// <exception cref="IOException"></exception>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public override int Read(byte[] buffer, int offset, int count)
        {
            int n;

            if (IsClosed)
            {
                throw new IOException("Stream is closed");
            }

            do
            {
                while (_bufferIsEmpty || _normInstance.HasNextEvent(0, 0))
                {
                    ProcessEvent();

                    if (_receivedEof) 
                    { 
                        return -1; 
                    }
                }

                if (_normStream == null)
                {
                    return -1;
                }

                // Read from the stream
                if ((n = _normStream.Read(buffer, offset, count)) < 0)
                {
                    throw new IOException("Break in stream integrity");
                }

                _bufferIsEmpty = n == 0;
            } 
            while (_bufferIsEmpty);

            return n;
        }

        private void ProcessEvent()
        {
            // Retrieve the next event
            var normEvent = _normInstance.GetNextEvent();

            // Check if the stream was closed
            if (IsClosed)
            {
                throw new IOException("Stream closed");
            }

            if (normEvent != null)
            {
                // Process the event
                var eventType = normEvent.Type;
                switch (eventType)
                {
                    case NormEventType.NORM_RX_OBJECT_NEW:
                        var normObject = normEvent.Object;
                        if (normObject != null && normObject.Type == NormObjectType.NORM_OBJECT_STREAM)
                        {
                            _normStream = (NormStream)normObject;
                        }
                        break;

                    case NormEventType.NORM_RX_OBJECT_UPDATED:
                        var theNormObject = normEvent.Object;
                        if (theNormObject == null || !theNormObject.Equals(_normStream))
                        {
                            break;
                        }

                        // Signal that the buffer is not empty
                        _bufferIsEmpty = false;
                        break;

                    case NormEventType.NORM_RX_OBJECT_ABORTED:
                    case NormEventType.NORM_RX_OBJECT_COMPLETED:
                        _normStream = null;

                        // Signal that the stream has ended
                        _receivedEof = true;
                        break;

                    default:
                        break;
                }

                // Notify listeners of the norm event
                FireNormEventOccured(normEvent);
            }
        }

        public override void Flush()
        {
            throw new NotSupportedException();
        }

        public override void Write(byte[] buffer, int offset, int count)
        {
            throw new NotSupportedException();
        }

        public override long Seek(long offset, SeekOrigin origin)
        {
            throw new NotSupportedException();
        }

        public override void SetLength(long value)
        {
            throw new NotSupportedException();
        }

        public override bool CanRead => true;

        public override bool CanSeek => false;

        public override bool CanWrite => false;

        public override long Length => throw new NotSupportedException();

        public override long Position { get => throw new NotSupportedException(); set => throw new NotSupportedException(); }
    }
}